### PR TITLE
feat: implement key storage

### PR DESCRIPTION
**Pull request by DmitriyMV**
This commits adds a new type - keystorage.KeyStorage which is used to safely store master key.

For #131

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

### CODE REVIEW

File: pkg/keystorage/keystorage.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+// KeyStorage is a key storage that can be used to store and retrieve the master key.
+//
+//nolint:govet
+type KeyStorage struct {
+	mx         sync.Mutex
+	underlying key_storage.Storage
+}
+
+// SetMasterKey sets the master key for the key storage, encrypts it using public key and stores it in slot id.
+func (ks *KeyStorage) SetMasterKey(masterKey []byte, slotID, slotPublicKey string) error {
```
Comment (smira) [inline pkg/keystorage/keystorage.go:29]: suggestion: we could call this `Initialize()` (or something similar) and make this function generate the random `masterKey` itself

File: pkg/keystorage/keystorage.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		EncryptedKey: []byte(encryptedSlot),
+	}
+
+	ks.underlying.KeysHmacHash = ks.hashSlots(masterKey)
+
+	return nil
+}
+
+// GetKey returns the attempts to decrypt the key slot with the given private key and return the master key.
+func (ks *KeyStorage) GetKey(slotID, slotPrivateKey string) ([]byte, error) {
```
Comment (smira) [inline pkg/keystorage/keystorage.go:102]: I think this returns the master key, to distinguish from `SetKey`, which operates on key slots

```suggestion
func (ks *KeyStorage) GetMasterKey(slotID, slotPrivateKey string) ([]byte, error) {
```

File: pkg/keystorage/keystorage.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	ks.underlying.KeysHmacHash = ks.hashSlots(masterKey)
+
+	return nil
+}
+
+// SetKey sets the new key slot with the given slot id and public key using previus slot and its private key.
+// It cannot be used to update an existing key slot.
+// It is required to call SetMasterKey() or UnmarshalBinary() to initialize the key storage first.
+func (ks *KeyStorage) SetKey(newSlotID, newSlotPublicKey, oldSlotID, oldSlotPrivateKey string) error {
```
Comment (smira) [inline pkg/keystorage/keystorage.go:66]: we probably need `DeleteSlot` function as well

and not sure about proper naming here, `EncryptSlot` maybe?

**Review by smira — APPROVED** [review]
my concern is that this doesn't include actual writes/reads to the database, so it might be hard to integrate this with the conflict check if it's done outside of this type -> e.g. if adding the key results in a conflict, the change should be reverted?

File: pkg/keystorage/keystorage.go (hunk 1 above)
Comment (DmitriyMV) [reply pkg/keystorage/keystorage.go]: Renamed to `Initialize` and added a new method `InitializeRnd`.

File: pkg/keystorage/keystorage.go (hunk 2 above)
Comment (DmitriyMV) [reply pkg/keystorage/keystorage.go]: Done!

File: pkg/keystorage/keystorage.go (hunk 3 above)
Comment (DmitriyMV) [reply pkg/keystorage/keystorage.go]: Changed to `AddKeySlot` and added `DeleteKeySlot`

**Pull request merged**